internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,205 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	tombpkg "gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tombpkg.Tomb) error {
+	var details checkDetails
+
+	m.state.Lock() // always acquire locks in same order (state lock, then plan lock)
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+	err := task.Get(checkDetailsAttr, &details)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for perform-check task %q: %v", task.ID(), err)
+	}
+	config, ok := plan.Checks[details.Name]
+	if !ok {
+		// Check no longer exists in plan.
+		return nil
+	}
+
+	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+	ticker := time.NewTicker(config.Period.Value)
+	defer ticker.Stop()
+
+	chk := newChecker(config, plan)
+	for {
+		select {
+		case <-ticker.C:
+			err := runCheck(tomb.Context(nil), chk, config.Timeout.Value)
+			if err != nil {
+				// Record check failure and perform any action if the threshold
+				// is reached (for example, restarting a service).
+				details.Failures++
+				m.updateHealthInfo(config, details.Failures)
+
+				m.state.Lock()
+				atThreshold := details.Failures >= config.Threshold
+				if atThreshold {
+					details.Proceed = true
+				} else {
+					// Add error to task log, but only if we haven't reached the
+					// threshold. When we hit the threshold, the "return err"
+					// below will cause the error to be logged.
+					logTaskError(task, err)
+				}
+				task.Set(checkDetailsAttr, &details)
+				m.state.Unlock()
+
+				logger.Noticef("Check %q failure %d/%d: %v", config.Name, details.Failures, config.Threshold, err)
+				if atThreshold {
+					logger.Noticef("Check %q threshold %d hit, triggering action and recovering", config.Name, config.Threshold)
+					m.callFailureHandlers(config.Name)
+					// Returning the error means perform-check goes to Error status
+					// and logs the error to the task log.
+					return err
+				}
+			} else if details.Failures > 0 {
+				m.updateHealthInfo(config, 0)
+
+				m.state.Lock()
+				task.Logf("succeeded after %s", pluralise(details.Failures, "failure", "failures"))
+				details.Failures = 0
+				task.Set(checkDetailsAttr, &details)
+				m.state.Unlock()
+			}
+
+		case <-tomb.Dying():
+			break
+		}
```

Comment:
If you ```break``` with a label to get outside the for loop, would the exit code below still be needed ?

Reply:
Yeah, because that `tombErr != tombpkg.ErrStillAlive` block (I assume that's what you mean by "exit code") is also needed in the case when we get a `ticker.C` and the context is cancelled (change aborted). `runCheck` returns nil in this case (see it's code) because the check itself didn't fail, but we still need to check the `tomb.Err()` and return if we're done. I thought a bit about a better way to structure this, but couldn't see an obvious way -- open to suggestions.